packages/core/src/tools/sanitize.ts

```diff
@@ -0,0 +1,242 @@
+import type { Context, ContextArray, ContextValue } from './context'
+import { display } from './display'
+import type { ObjectWithToJsonMethod } from './utils'
+import { detachToJsonMethod, ONE_KIBI_BYTE } from './utils'
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+type PrimitivesAndFunctions = string | number | boolean | undefined | null | symbol | bigint | Function
+type ExtendedContextValue = PrimitivesAndFunctions | object | ExtendedContext | ExtendedContextArray
+type ExtendedContext = { [key: string]: ExtendedContextValue }
+type ExtendedContextArray = ExtendedContextValue[]
+
+type ContainerElementToProcess = {
+  source: ExtendedContextArray | ExtendedContext
+  target: ContextArray | Context
+  path: string
+}
+
+// The maximum size of a single event is 256KiB. By default, we ensure that user-provided data
+// going through sanitize fits inside our events, while leaving room for other contexts, metadata, ...
+const SANITIZE_DEFAULT_MAX_CHARACTER_COUNT = 220 * ONE_KIBI_BYTE
+
+// Symbol for the root element of the JSONPath used for visited objects
+const JSON_PATH_ROOT_ELEMENT = '$'
+
+// When serializing (using JSON.stringify) a key of an object, { key: 42 } gets wrapped in quotes as "key".
+// With the separator (:), we need to add 3 characters to the count.
+const KEY_DECORATION_LENGTH = 3
+
+/**
+ * Ensures user-provided data is 'safe' for the SDK
+ * - Deep clones data
+ * - Removes cyclic references
+ * - Transforms unserializable types to a string representation
+ *
+ * LIMITATIONS:
+ * - Size is in characters, not byte count (may differ according to character encoding)
+ * - Size does not take into account indentation that can be applied to JSON.stringify
+ * - Non-numerical properties of Arrays are ignored. Same behavior as JSON.stringify
+ *
+ * @param source              User-provided data meant to be serialized using JSON.stringify
+ * @param maxCharacterCount   Maximum number of characters allowed in serialized form
+ */
+export function sanitize(source: string, maxCharacterCount?: number): string | undefined
+export function sanitize(source: Context, maxCharacterCount?: number): Context
+export function sanitize(source: unknown, maxCharacterCount?: number): ContextValue
+export function sanitize(source: unknown, maxCharacterCount = SANITIZE_DEFAULT_MAX_CHARACTER_COUNT) {
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏ Why not to put the feature flag check direcly in the `sanitize` function
```suggestion
export function sanitize(source: unknown, maxCharacterCount = SANITIZE_DEFAULT_MAX_CHARACTER_COUNT) {
if (!isExperimentalFeatureEnabled('sanitize_inputs')) {
  return source
}
```

Reply:
Sanitize is used in two cases:
- It replaces deepClone - so we'd need to return deepClone(source) when the featureFlag is not enabled
- It has been added to all user-provided data, including primitive types, which are not being routed through deepClone today.